typing/scalar.ml

```diff
@@ -0,0 +1,694 @@
+        | Div Safe -> "div"
+        | Div Unsafe -> "unsafe_div"
+        | Mod Safe -> "mod"
+        | Mod Unsafe -> "unsafe_mod"
+        | And -> "and"
+        | Or -> "or"
+        | Xor -> "xor"
+    end
+
+    module Shift_op = struct
+      module Rhs = struct
+        (* CR jvanburen: expand this to all of [Integral.t] *)
+        type t = Int
+
+        let all = [Int]
+      end
+
+      type t =
+        | Lsl
+        | Asr
+        | Lsr
+
+      let all = [Lsl; Asr; Lsr]
+
+      let to_string = function Lsl -> "lsl" | Asr -> "asr" | Lsr -> "lsr"
+    end
+
+    module Float_op = struct
+      type t =
+        | Add
+        | Sub
+        | Mul
+        | Div
+
+      let all = [Add; Sub; Mul; Div]
+
+      let to_string = function
+        | Add -> "add"
+        | Sub -> "sub"
+        | Mul -> "mul"
+        | Div -> "div"
+    end
+
+    (* CR jvanburen: comparisons that return naked values *)
+
+    (** comparisons return a tagged immediate *)
+    type nonrec 'mode t =
+      (* CR jvanburen: Fmod, Min, Max? *)
+      | Integral of 'mode Integral.t * Int_op.t
+      | Shift of 'mode Integral.t * Shift_op.t * Shift_op.Rhs.t
+      | Floating of 'mode Floating.t * Float_op.t
+      | Icmp of any_locality_mode Integral.t * Integer_comparison.t
+      | Fcmp of any_locality_mode Floating.t * Float_comparison.t
+      | Three_way_compare of any_locality_mode t
+
+    let all =
+      List.concat
+        [ ListLabels.concat_map Integral.all ~f:(fun size ->
+              ListLabels.map Int_op.all ~f:(fun op -> Integral (size, op)));
+          ListLabels.concat_map Floating.all ~f:(fun size ->
+              ListLabels.map Float_op.all ~f:(fun op -> Floating (size, op)));
+          ListLabels.concat_map Integral.all ~f:(fun size ->
+              ListLabels.concat_map Shift_op.all ~f:(fun op ->
+                  ListLabels.map Shift_op.Rhs.all ~f:(fun rhs ->
+                      Shift (size, op, rhs))));
+          ListLabels.concat_map Integer_comparison.all ~f:(fun cmp ->
+              List.map (fun size -> Icmp (size, cmp)) Integral.all);
+          ListLabels.concat_map Float_comparison.all ~f:(fun cmp ->
+              List.map (fun size -> Fcmp (size, cmp)) Floating.all);
+          List.map (fun size -> Three_way_compare size) all ]
+
+    let sort = function
+      | Integral
+          ( width,
+            ( Add | Sub | Mul
+            | Div (Safe | Unsafe)
+            | Mod (Safe | Unsafe)
+            | And | Or | Xor ) ) ->
+        let sort = Integral.sort width in
+        sort, sort, sort
+      | Shift (width, (Lsl | Lsr | Asr), Int) ->
+        let sort = Integral.sort width in
+        sort, Jkind_types.Sort.Const.value, sort
+      | Floating (width, (Add | Sub | Mul | Div)) ->
+        let sort = Floating.sort width in
+        sort, sort, sort
+      | Icmp (width, (_ : Integer_comparison.t)) ->
+        let sort = Integral.sort width in
+        sort, sort, Jkind_types.Sort.Const.value
```

Comment:
Similarly, I would expect things like this to be able to produce untagged booleans.